### PR TITLE
Support multiple uid

### PR DESCRIPTION
**Pull request by live1206**
#3457 
- Users should be able to define multiple uids within one SDP
- For uid defined in non-root level, the related href should append bookmark converted form uid

### CODE REVIEW

File: src/docfx/build/xref/XrefMap.cs [hunk 1]
```diff
@@ -354,21 +358,105 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
                 obj,
                 schema.Type,
                 transform: AttributeTransformer.TransformXref(context, file, null, extensionData));
 
             errors.AddRange(schemaErrors);
-            var xref = new InternalXrefSpec
+
+            var extensionDataByUid = new Dictionary<string, (bool isRoot, Dictionary<string, Lazy<JValue>> properties)>();
+
+            if (extensionData.Count == 0)
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: nit: no need to special case `extensionData.Count == 0` here, can just loop over it.

File: src/docfx/build/xref/XrefMap.cs [hunk 2]
```diff
@@ -354,21 +358,105 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
+                    Href = isRoot ? file.CanonicalUrlWithoutLocale : $"{file.CanonicalUrlWithoutLocale}#{GetBookmarkFromUid(item.Key)}",
+                    ReferencedFile = file,
+                };
+                xref.ExtensionData.AddRange(properties);
+                return xref;
+            }).ToList();
+
+            return (errors, specs);
+
+            string GetBookmarkFromUid(string uid)
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: different uids may map to the same bookmark using this algorithm, how about using `Uri.EscapeDataString`?

File: src/docfx/build/xref/XrefMap.cs [hunk 3]
```diff
@@ -354,21 +358,105 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
+                => Regex.Replace(uid, @"\W", "_");
+
+            (string uid, string jsonPath) MatchExtensionDataToUid(string jsonPath)
+            {
+                var reverse = uids.ToDictionary(x => x.Value, x => x.Key);
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: build the `reverse` lookup upfront to save duplicated work here. Name them something like `uidToJsonPath` and `jsonPathToUid` so we know what each `Dict<string, string>` represents

File: src/docfx/build/xref/XrefMap.cs [hunk 4]
```diff
@@ -354,21 +358,105 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
+                    context.Report.Write(filePath, Errors.UidConflict(str));
+                }
+            }
+
+            foreach (var item in token.Children())
+            {
+                var property = item as JProperty;
+                if (property.Value is JObject obj)
+                {
+                    GetUids(context, filePath, obj, string.IsNullOrEmpty(parentName) ? property.Name : $"{parentName}.{property.Name}", uids);
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: use `JToken.Path` over manually building json path

File: src/docfx/build/xref/XrefMap.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: This logic is from the template of v2, just want to follow the same behavior

File: src/docfx/build/xref/XrefMap.cs (hunk 3 above)
Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: sounds good

File: src/docfx/build/xref/XrefMap.cs [hunk 5]
```diff
@@ -354,21 +359,102 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
+        }
+
+        private static void GetUids(Context context, string filePath, JObject token, string parentName, Dictionary<string, string> uidToJsonPath, Dictionary<string, string> jsonPathToUid)
+        {
+            if (token is null)
+                return;
+
+            if (token.TryGetValue("uid", out var value) && value is JValue v && v.Value is string str)
+            {
+                if (!uidToJsonPath.TryAdd(str, parentName))
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: is this `token.Path`, if so then `parentName` is not needed.

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: u r right 👍

File: docs/specs/xref.yml [hunk 6]
```diff
@@ -222,7 +222,56 @@ outputs:
...
 ---
-# Define uid in non-root object level, it will be ignored
+# support multiple uids in the same array in one SDP
+inputs:
+  docfx.yml:
+  docs/a.yml: |
+    #YamlMime:TestData
+    uid: a
+    description: description for a
+    array:
```
Comment (OsmondJiang) [inline docs/specs/xref.yml]: add test for single `data`

File: src/docfx/build/xref/XrefMap.cs [hunk 7]
```diff
@@ -354,21 +359,102 @@ private static (List<Error> errors, InternalXrefSpec spec) LoadSchemaDocument(Co
...
+        private static void GetUids(Context context, string filePath, JObject token, Dictionary<string, string> uidToJsonPath, Dictionary<string, string> jsonPathToUid)
+        {
+            if (token is null)
+                return;
+
+            if (token.TryGetValue("uid", out var value) && value is JValue v && v.Value is string str)
+            {
+                if (!uidToJsonPath.TryAdd(str, token.Path))
+                {
+                    context.Report.Write(filePath, Errors.UidConflict(str));
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: >context [](start = 20, length = 7)

does it only woking for inside conflicts?

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: Here we only want to detect duplicated uids defined in one file.
The case missing is, detect duplicated uid defined already externally, defined again in the build repo

Comment (OsmondJiang) [reply src/docfx/build/xref/XrefMap.cs]: no need comparing with external uids, but make sure the uid is unique inside this docset. add test for that

---
In reply to: [250914849](https://github.com/dotnet/docfx/pull/3934#discussion_r250914849) [](ancestors = 250914849)

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: test added

**Pull request merged**